### PR TITLE
Add JsonExtensionData to ProblemDetails

### LINKED ISSUE #6202: Figure out ProblemDetails + System.Text.Json
* [ ] - `ProblemDetails` relied on `JsonExtensionAttribute` to ensure that swagger document generators produced the correct JsonSchema. Since the attribute would no longer be available, we need  to figure out an alternate way to describe this to document generators.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/aspnet/AspNetCore/issues/6202

### CODE REVIEW

File: src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs [hunk 1]
```diff
@@ -59,7 +59,12 @@ public sealed override async Task WriteResponseBodyAsync(OutputFormatterWriteCon
             var writeStream = GetWriteStream(httpContext, selectedEncoding);
             try
             {
-                await JsonSerializer.WriteAsync(writeStream, context.Object, context.ObjectType, SerializerOptions);
+                // context.ObjectType reflects the declared model type when specified.
+                // For polymorphic scenarios where the user declares a return type, but returns a derived type,
+                // we want to serialize all the properties on the derived type. This keeps parity with
+                // the behavior you get when the user does not declare the return type and with Json.Net at least at the top level.
+                var objectType = context.Object?.GetType() ?? context.ObjectType;
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs]: What do you think? I considered leaving it as is, but you get a different outcome depending on whether you do `ActionResult<T>` versus returning an `ObjectResult`. Doing this felt like it was moderately better. Polymorphic serialization does not work for properties, but I think there's plans for a feature there.

File: src/Mvc/test/Mvc.FunctionalTests/ApiBehaviorTest.cs [hunk 2]
```diff
@@ -364,4 +332,132 @@ public async Task SerializingValidationProblemDetails_WithExtensionData()
...
+                    var error = Assert.Single(kvp.Value);
+                    Assert.Equal("The field Zip must match the regular expression '\\d{5}'.", error);
+                }
+            );
+
+            Assert.Collection(
+                problemDetails.Extensions,
+                 kvp =>
+                 {
+                     Assert.Equal("extensions", kvp.Key);
```
Comment (pranavkm) [inline src/Mvc/test/Mvc.FunctionalTests/ApiBehaviorTest.cs]: This is kinda busted - https://github.com/dotnet/corefx/issues/38769. I've added a test to make sure the rest of Json is correct. If we get this in, the most we'd have to do to "react" is to fix up tests

File: src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs (hunk 1 above)
Comment (rynowak) [reply src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs]: I think this is the right thing to do given that JSON.net does it.

File: src/Mvc/test/Mvc.FunctionalTests/ApiBehaviorTest.cs (hunk 2 above)
Comment (rynowak) [reply src/Mvc/test/Mvc.FunctionalTests/ApiBehaviorTest.cs]: Should we skip this test instead? I don't really want to encode wrong behavior in tests. It's better to just skip if possible.

Put an other way, I don't really want a test that partially validates the expectations.

**Review by rynowak — APPROVED** [review]

File: src/Mvc/test/Mvc.FunctionalTests/ApiBehaviorTest.cs (hunk 2 above)
Comment (pranavkm) [reply src/Mvc/test/Mvc.FunctionalTests/ApiBehaviorTest.cs]: Yeah, seems reasonable. It's likely https://github.com/dotnet/corefx/issues/38769 will get fixed for preview7.

**Pull request merged**